### PR TITLE
Introduce GoogleTest framework for Valkey unit testing

### PR DESCRIPTION
**Pull request by harrylin98**
### This PR adds GoogleTest (gtest) support to Valkey to enable writing modern unit tests.

**Motivation**: 
GoogleTest provides richer assertions, test fixtures, mocking support, and improved diagnostics, helping improve test coverage and maintainability over time.

**Summary**:
The change is limited to test infrastructure, and existing C unit tests remain unchanged.

This PR focuses only on integrating the framework and includes a small set of example tests to demonstrate usage.

For more details, see `src/gtest/README.md`.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -55,6 +58,11 @@ jobs:
           - {version: "8.1.3", file: "valkey-8.1.3-noble-x86_64.tar.gz"}
     steps:
       - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      - name: Install modern CMake
```
Comment (madolson) [inline .github/workflows/ci.yml:61]: This builds with normal make, we shouldn't need cmake here.

**Comment by madolson** [comment]
Some high level things:
1. Let's try to merge this into the forkless branch, it should be merged into unstable. 
2. I don't think we should have two unit testing frameworks. I feel like we need to commit to incorporating the google testing framework. I'm not saying that is a blocker to start, but we should agree to move what we have.
3. We typically deep copy dependencies to make vending easier, maybe we don't care in this case because it is optional and not required? Not a particularly strong opinion.

File: src/gtest/example_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// Test matcher works in custom_matchers.hpp
+TEST_F(ExampleTest, TestMatchers) {
+    robj *robj_str = createStringObject("test", 4);
+    ASSERT_NE(robj_str , nullptr); // "ASSERT" is correct here, because the test can't reasonably continue
+    EXPECT_THAT(robj_str, robjEqualsStr("test"));
+    decrRefCount(robj_str);
+}
+
+// Verify mocking works via zfree
+TEST_F(ExampleTest, TestMocking) {
```
Comment (madolson) [inline src/gtest/example_tests.cpp:61]: I think this is the type of mocking that got ElastiCache into trouble. This is trying to make sure zfree is called, but there are better systemic ways like https://github.com/valkey-io/valkey/blob/0a9cb985c8c37c54c5007bc0af2dcd0b6752a529/src/server.h#L4077.

File: src/gtest/wrappers.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * The list of wrapper methods defined.  Each wrapper method must
+ * conform to the same naming conventions (i.e. prefix with a
+ * '__wrap_') and have its method signature match the overridden
+ * method exactly.
+ *
+ * Note: You should NOT wrap variable argument functions (i.e have "...")
+ *       See: https://github.com/google/googletest/blob/master/googlemock/docs/gmock_faq.md#can-i-mock-a-variadic-function
+ *       Example: serverLog(int level, const char *fmt, ...) should NOT be mocked.
+ */
+long long __wrap_aeCreateTimeEvent(aeEventLoop *eventLoop, long long milliseconds, aeTimeProc *proc, void *clientData, aeEventFinalizerProc *finalizerProc);
```
Comment (madolson) [inline src/gtest/wrappers.h:56]: I think this is the only valid wrap in this list. I don't think any of these other functions should be wrapped. I would like to see up primarily wrap functions that have non-deterministic output. (Network, timers, disk, etc)

File: .github/workflows/ci.yml (hunk 1 above)
Comment (harrylin98) [reply .github/workflows/ci.yml:61]: We need to build the GoogleTest dependencies, which require CMake. Check gtest target in `deps/Makefile`.

File: src/gtest/README.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+- Integrated mocking via GoogleMock, providing expressive, behavior-based mocks
+- Rich argument matchers and call sequencing / ordering
+- Advanced test fixtures and lifecycle control
+- Richful assertions and detailed diagnostics
+- Parameterized and typed tests
+
+These features enable more expressive, maintainable, and scalable unit tests,
+particularly for complex components and edge-case validation.
+
+For more information on GoogleTest, see: https://google.github.io/googletest/
```
Comment (madolson) [inline src/gtest/README.md:17]:
```suggestion
```
Put this in the PR, someone looking for docs doesn't need this information.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (madolson) [reply .github/workflows/ci.yml:61]: Requiring CMake is a bit of annoying dependency to be taking.
